mithril-aggregator/src/http_server/routes/artifact_routes/cardano_transaction.rs

```diff
@@ -0,0 +1,302 @@
+    use crate::services::MessageService;
+
+    use slog_scope::{debug, warn};
+    use std::convert::Infallible;
+    use std::sync::Arc;
+    use warp::http::StatusCode;
+
+    pub const LIST_MAX_ITEMS: usize = 20;
+
+    /// List Cardano Transactions set artifacts
+    pub async fn list_artifacts(
+        http_message_service: Arc<dyn MessageService>,
+    ) -> Result<impl warp::Reply, Infallible> {
+        debug!("⇄ HTTP SERVER: artifacts");
+
+        match http_message_service
+            .get_cardano_transaction_list_message(LIST_MAX_ITEMS)
+            .await
+        {
+            Ok(message) => Ok(reply::json(&message, StatusCode::OK)),
+            Err(err) => {
+                warn!("list_artifacts_cardano_transactions"; "error" => ?err);
+
+                Ok(reply::internal_server_error(err))
+            }
+        }
+    }
+
+    /// Get Artifact by signed entity id
+    pub async fn get_artifact_by_signed_entity_id(
+        signed_entity_id: String,
+        http_message_service: Arc<dyn MessageService>,
+    ) -> Result<impl warp::Reply, Infallible> {
+        debug!("⇄ HTTP SERVER: artifact/{signed_entity_id}");
+
+        match http_message_service
+            .get_cardano_transaction_message(&signed_entity_id)
+            .await
+        {
+            Ok(Some(message)) => Ok(reply::json(&message, StatusCode::OK)),
+            Ok(None) => {
+                warn!("get_cardano_transaction_details::not_found");
+                Ok(reply::empty(StatusCode::NOT_FOUND))
+            }
+            Err(err) => {
+                warn!("get_cardano_transaction_details::error"; "error" => ?err);
+                Ok(reply::internal_server_error(err))
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+pub mod tests {
+    use crate::{
+        http_server::SERVER_BASE_PATH,
+        initialize_dependencies,
+        message_adapters::{
+            ToCardanoTransactionListMessageAdapter, ToCardanoTransactionMessageAdapter,
+        },
+        services::MockMessageService,
+    };
+    use chrono::{DateTime, Utc};
+    use mithril_common::{
+        entities::{Beacon, SignedEntity, SignedEntityType},
+        messages::ToMessageAdapter,
+        signable_builder::Artifact,
+        sqlite::HydrationError,
+        test_utils::{apispec::APISpec, fake_data},
+    };
+    use serde_json::Value::Null;
+    use warp::{http::Method, test::request};
+
+    use super::*;
+
+    fn setup_router(
+        dependency_manager: Arc<DependencyContainer>,
+    ) -> impl Filter<Extract = (impl warp::Reply,), Error = warp::Rejection> + Clone {
+        let cors = warp::cors()
+            .allow_any_origin()
+            .allow_headers(vec!["content-type"])
+            .allow_methods(vec![Method::GET, Method::POST, Method::OPTIONS]);
+
+        warp::any()
+            .and(warp::path(SERVER_BASE_PATH))
+            .and(routes(dependency_manager).with(cors))
+    }
+
+    pub fn create_signed_entities<T>(
```

Comment:
This is the second time this method is copied in this module, maybe we could declare it in the module directly behind a `#[cfg(test)]` flag ?

Reply:
It looks like it's not that easy as the `setup_router` function uses the `routes` function of the sub-modules.
It got complicated very quickly and I couldn't even make it compile because of limitations of trait objects.
Let's keep it like this and give it another try together later :slightly_smiling_face: